### PR TITLE
feat(linter): add no-bitwise rule

### PR DESCRIPTION
**Pull request by mysteryven**
feat: #227

### CODE REVIEW

File: crates/oxc_ast/src/ast/operator.rs [hunk 1]
```diff
@@ -98,7 +98,15 @@ impl AssignmentOperator {
 
     #[must_use]
     pub fn is_bitwise(self) -> bool {
-        matches!(self, Self::BitwiseOR | Self::BitwiseXOR | Self::BitwiseAnd)
+        matches!(
+            self,
+            Self::BitwiseOR
+                | Self::BitwiseXOR
+                | Self::BitwiseAnd
+                | Self::ShiftLeft
```
Comment (mysteryven) [inline crates/oxc_ast/src/ast/operator.rs:106]: Eslint has those [test suits](https://github.com/eslint/eslint/blob/e39f28d8578a00f4da8d4ddad559547950128a0d/tests/lib/rules/no-bitwise.js#L47-L49), so I add more bitwises for Assignment.

File: crates/oxc_linter/src/rules/no_bitwise.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    use crate::tester::Tester;
+
+    let pass = vec![
+        ("a + b", None),
+        ("!a", None),
+        ("a && b", None),
+        ("a || b", None),
+        ("a += b", None),
+        ("a &&= b", None),
```
Comment (mysteryven) [inline crates/oxc_linter/src/rules/no_bitwise.rs:134]: ESlint add [`parserOptions`](https://github.com/eslint/eslint/blob/e39f28d8578a00f4da8d4ddad559547950128a0d/tests/lib/rules/no-bitwise.js#L28-L30) for those, I not sure how we can add it, or does it needed for Oxc?

File: crates/oxc_linter/src/rules/no_bitwise.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        ("a ||= b", None),
+        ("a ??= b", None),
```
Comment (mysteryven) [inline crates/oxc_linter/src/rules/no_bitwise.rs:136]: ESLint add [parserOptions](https://github.com/eslint/eslint/blob/e39f28d8578a00f4da8d4ddad559547950128a0d/tests/lib/rules/no-bitwise.js#L28-L30) for above three cases, I not sure how we can add it, or does it needed for Oxc?

```js
 { code: "a &&= b", parserOptions: { ecmaVersion: 2021 } },
 { code: "a ||= b", parserOptions: { ecmaVersion: 2021 } },
 { code: "a ??= b", parserOptions: { ecmaVersion: 2021 } },
```

**Comment by Boshen** [comment]
It seems we have misunderstanding here,

[`bad-bitwise-operator`](https://deepscan.io/docs/rules/bad-bitwise-operator) is from deepscan, where it forbids bitwise in certain constructs.
[`no-bitwsise`](https://eslint.org/docs/latest/rules/no-bitwise) is for forbidding all bitwise operations.

This is why `bad-bitwise-operator` can be categoriezed as correctness. And `no-bitwsise` is not a recommended rule in ESLint.

File: crates/oxc_linter/src/rules/no_bitwise.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    /// The use of bitwise operators in JavaScript is very rare and often `&` or `|` is simply a mistyped `&&` or `||`,
+    /// which will lead to unexpected behavior.
+    ///
+    /// ### Example
+    ///
+    /// ```javascript
+    /// var x = y | z;
+    /// ```
+    NoBitwise,
+    correctness
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_bitwise.rs:37]: You can change this to nursery.

File: crates/oxc_linter/src/rules/no_bitwise.rs (hunk 3 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/no_bitwise.rs:136]: We don't need parser options, we parse everything!

File: crates/oxc_linter/src/snapshots/no_bitwise.snap [hunk 5]
```diff
@@ -0,0 +1,83 @@
+---
+source: crates/oxc_linter/src/tester.rs
+expression: no_bitwise
+---
+
+  ⚠ eslint(no-bitwise): Unexpected use of "^"
+   ╭─[no_bitwise.tsx:1:1]
+ 1 │ a ^ b
+   · ─────
```
Comment (Boshen) [inline crates/oxc_linter/src/snapshots/no_bitwise.snap:9]: It's best to point to the operator, but we don't have a token 🤔 Are we able to get. the source and I find it 🤔

File: crates/oxc_linter/src/rules/no_bitwise.rs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-bitwise): Unexpected use of {0:?}")]
+#[diagnostic(severity(warning))]
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_bitwise.rs:12]: It would be the best if we add some help here, why is it unexpected?

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_ast/src/ast/operator.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_ast/src/ast/operator.rs:106]: I fixed this on main.

File: crates/oxc_linter/src/snapshots/no_bitwise.snap (hunk 5 above)
Comment (mysteryven) [reply crates/oxc_linter/src/snapshots/no_bitwise.snap:9]: I am stuck on how to compute the left expression's length in nested expression. e.g. 

```js
(a | b) 
// ^ this is the expected position

(a | b) | c
    //  ^ this is the expected position
```

Comment (Boshen) [reply crates/oxc_linter/src/snapshots/no_bitwise.snap:9]: I think it's impossible for a AST, let's give up 😄

File: crates/oxc_linter/src/rules/no_bitwise.rs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+fn is_int32_hint(int32_hint: bool, node: &AstNode) -> bool {
+    if !int32_hint {
+        return false;
+    }
+
+    match node.get().kind() {
+        AstKind::BinaryExpression(bin_expr) => {
+            bin_expr.operator.as_str() == "|" && bin_expr.right.is_number_0()
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_bitwise.rs:113]: You can match on the operator: `bin_expr.operator == BinaryOperator::BitwiseOR`.
We are not JavaScript ;-) we don't need to match on strings :-)

**Review by Boshen — CHANGES_REQUESTED** [review]
Last small change and we'll be able to merge.

File: crates/oxc_linter/src/rules/no_bitwise.rs (hunk 7 above)
Comment (mysteryven) [reply crates/oxc_linter/src/rules/no_bitwise.rs]: done 😀

**Comment by Boshen** [comment]
Thank you so much!

**Pull request merged**